### PR TITLE
[Wallet] Improve rescan API + avoid permanent locks during RescanFromTime

### PR DESCRIPTION
**Pull request by random-zebra**
Based on top of:
- [x] #2279 

Backports bitcoin#10412, and bitcoin#11281

### CODE REVIEW

**Comment by random-zebra** [comment]
Rebased on master. Ready to go.

**Review by furszy — COMMENTED** [review]
Initial quick code review ACK 29c267ef2bd367ec6218e2b0ad2d7d7e2571e7b1

Overall is looking really nice 👌

**Review by Fuzzbawls — COMMENTED** [review]
Functional ACK

Great improvement in terms of the wallet's responsiveness! have been able to import/rescan addresses and wallets with thousands of transactions without issue.

One commented change in the `importmulti` help output that shouldn't be, otherwise good to go

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -1037,10 +1037,10 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
                                                           "block from time %d, which is after or within %d seconds of key creation, and "
                                                           "could contain transactions pertaining to the key. As a result, transactions "
                                                           "and coins using this key may not appear in the wallet. This error could be "
-                                                          "caused by pruning or data corruption (see pivxd log for details) and could "
+                                                          "caused by pruning or data corruption (see bitcoind log for details) and could "
```
[the remark this answers is not in the document]
Comment (random-zebra) [reply src/wallet/rpcdump.cpp]: fixed :+1:

**Review by Fuzzbawls — APPROVED** [review]
ACK 29c267ef2bd367ec6218e2b0ad2d7d7e2571e7b1

**Review by furszy — APPROVED** [review]
ACK 1cde8b4 and merging

**Pull request merged**